File: src/reserve-stats.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { initial } from "./game.js";
import { startInjury,medicalStep,pickup } from "./features.js";
test("full energy restores the current substitute's saved stats through a full bench rotation", () => {
 const s=initial(),p=s.players[7];
 s.bench[0]=[170,140,120].map(v=>Array(8).fill(v));p.stats.fill(173);
 for(const expected of [170,140,120,170]) {
  Object.assign(p,{x:0,z:0,health:0,actionTime:0});assert.ok(startInjury(s,7));
  for(let i=0;i<1000&&s.medical;i++)medicalStep(s,.04);
  assert.equal(s.medical,null);
  assert.deepEqual(p.baseStats,Array(8).fill(expected));
  p.stats.fill(100);p.health=20;pickup(s,7,11);
  assert.equal(p.health,100);assert.deepEqual(p.stats,Array(8).fill(expected));
  assert.notEqual(p.stats,p.baseStats);
 }
});

File: server/reserve_stats_test.go
package main

import "testing"

func TestReserveFullEnergyBaseline(t *testing.T) {
	s := initial()
	p := &s.Players[7]
	for i, v := range []int{170, 140, 120} {
		for j := range s.Bench[0][i] {
			s.Bench[0][i][j] = v
		}
	}
	for j := range p.Stats {
		p.Stats[j] = 173
	}
	for _, expected := range []int{170, 140, 120, 170} {
		p.X = 0
		p.Z = 0
		p.Health = 0
		p.ActionTime = 0
		if !s.startInjury(7) {
			t.Fatal("injury start")
		}
		for i := 0; i < 1000 && s.Medical != nil; i++ {
			s.medicalStep(.04)
		}
		if s.Medical != nil {
			t.Fatal("transport incomplete")
		}
		for _, v := range p.BaseStats {
			if v != expected {
				t.Fatal("substitute baseline", expected, v)
			}
		}
		p.Stats = defaultStats()
		p.Health = 20
		s.pickup(7, 11)
		if p.Health != 100 {
			t.Fatal("energy")
		}
		for _, v := range p.Stats {
			if v != expected {
				t.Fatal("restored stats", expected, v)
			}
		}
	}
}
